node/src/exchange/client.rs

```diff
@@ -56,63 +96,697 @@ impl ExchangeClientHandler {
+        let mut mock_req = MockReq::new();
+        let mut handler = ExchangeClientHandler::<MockReq>::new(peer_tracker);
+
+        let (tx, rx) = oneshot::channel();
+
+        handler.on_send_request(&mut mock_req, HeaderRequest::with_origin(0, 1), tx);
+
+        let expected = gen_header_response(10);
+        let expected_header = expected.to_extended_header().unwrap();
+
+        mock_req.send_n_responses(&mut handler, 1, vec![expected]);
+
+        for height in 1..10 {
+            mock_req.send_n_responses(&mut handler, 1, vec![gen_header_response(height)]);
+        }
+
+        let result = rx.await.unwrap().unwrap();
+        assert_eq!(result.len(), 1);
+        assert_eq!(result[0], expected_header);
+    }
+
+    #[tokio::test]
+    async fn head_request_responds_with_only_failures() {
+        let peer_tracker = peer_tracker_with_n_peers(15);
+        let mut mock_req = MockReq::new();
+        let mut handler = ExchangeClientHandler::<MockReq>::new(peer_tracker);
+
+        let (tx, rx) = oneshot::channel();
+
+        handler.on_send_request(&mut mock_req, HeaderRequest::with_origin(0, 1), tx);
+
+        mock_req.send_n_failures(&mut handler, 5, OutboundFailure::Timeout);
+        mock_req.send_n_failures(&mut handler, 5, OutboundFailure::ConnectionClosed);
+
+        assert!(matches!(
+            rx.await,
+            Ok(Err(P2pError::Exchange(ExchangeError::HeaderNotFound)))
+        ));
+    }
+
+    #[tokio::test]
+    async fn head_request_with_one_peer() {
+        let peer_tracker = peer_tracker_with_n_peers(1);
+        let mut mock_req = MockReq::new();
+        let mut handler = ExchangeClientHandler::<MockReq>::new(peer_tracker);
+
+        let (tx, rx) = oneshot::channel();
+
+        handler.on_send_request(&mut mock_req, HeaderRequest::with_origin(0, 1), tx);
+
+        let expected = gen_header_response(10);
+        let expected_header = expected.to_extended_header().unwrap();
+
+        mock_req.send_n_responses(&mut handler, 1, vec![expected]);
+
+        let result = rx.await.unwrap().unwrap();
+        assert_eq!(result.len(), 1);
+        assert_eq!(result[0], expected_header);
+    }
+
+    #[tokio::test]
+    async fn head_request_with_no_peers() {
+        let peer_tracker = peer_tracker_with_n_peers(0);
+        let mut mock_req = MockReq::new();
+        let mut handler = ExchangeClientHandler::<MockReq>::new(peer_tracker);
+
+        let (tx, rx) = oneshot::channel();
+
+        handler.on_send_request(&mut mock_req, HeaderRequest::with_origin(0, 1), tx);
+
+        assert!(matches!(rx.await, Ok(Err(P2pError::NoPeers))));
+    }
+
+    #[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+    struct MockReqId(u64);
+
+    impl MockReqId {
+        fn new() -> MockReqId {
+            static NEXT: AtomicU64 = AtomicU64::new(0);
+            let id = NEXT.fetch_add(1, Ordering::Relaxed);
+            MockReqId(id)
+        }
+    }
+
+    struct MockReq {
+        reqs: VecDeque<MockReqInfo>,
```

Comment:
I'm curious what's the point of `VecDeque`? is drain here more efficient?

Reply:
Yes. `Vec::drain` moves the remaining items at the beginning of the allocated memory.

Reply:
woow, that's a game changer, thanks!